content/solids/components/voxel-painter.ts

```diff
@@ -442,6 +494,29 @@ export class VoxelPainter extends CustomElementView {
     this.updateApplet();
   }
 
+  get voxelPositions(): [number, number, number][] {
+    return this.voxels.map(voxel => {
+      const pos = voxel.position;
+      return [pos.x, pos.y, pos.z];
+    });
+  }
+
+  onVolumeMet(fn: (positions: [number, number, number][]) => void) {
+    this.onTargetVolumeMet = fn;
+  }
+
+  onSurfaceMet(fn: (positions: [number, number, number][]) => void) {
+    this.onTargetSurfaceMet = fn;
+  }
+
+  onVolumeAreaMet(fn: (positions: [number, number, number][]) => void) {
+    this.onTargetVolumeAreaMet = fn;
+  }
+
+  onCameraSnap(fn: (side: Side) => void) {
+    this.onSnap = fn;
+  }
```

Comment:
Rather than setting these properties with callback functions, do something like

``` ts
this.trigger('volume-area-met', [1, 2, 3]);
```

And then listen for the event using

``` ts
$voxelPainter.on('volume-area-met', ([a, b, c]) => {
  // ...
});
```